rocks-lib/src/operations/install.rs

```diff
@@ -20,6 +20,81 @@ use thiserror::Error;
 
 use super::{resolve::get_all_dependencies, SearchAndDownloadError};
 
+/// A rocks package installer, providing fine-grained control
+/// over how packages should be installed.
+/// Can install multiple packages in parallel.
+pub struct Install<'a> {
+    config: &'a Config,
+    package_db: Option<RemotePackageDB>,
+    packages: Vec<(BuildBehaviour, PackageReq)>,
+    pin: PinnedState,
+    progress: Option<Arc<Progress<MultiProgress>>>,
+}
+
+impl<'a> Install<'a> {
+    /// Construct a new installer.
+    pub fn new(config: &'a Config) -> Self {
+        Self {
+            config,
+            package_db: None,
+            packages: Vec::new(),
```

Comment:
issue: i think the builder pattern doesn't quite fit in install's case since one can install an empty set of packages by simply running `Install::new().install()` (which makes little sense). Can we enforce at the type level that at least one package should be provided (without having to do runtime checks in `install()`)?

Reply:
This is also possible without the builder pattern, by passing in an empty vec.
We could change the `new` constructor to take a [`NonEmpty<PackageReq>`](https://docs.rs/nonempty/latest/nonempty/).

I'd be in favour of that, but that would be for a separate PR.

Reply:
Come to think of it, sometimes we want it to be possible for the list to be empty, for example when installing test dependencies.